src/main/java/io/vertx/ext/cluster/infinispan/impl/CloseableIteratorCollectionStream.java

```diff
@@ -0,0 +1,205 @@
+        synchronized (this) {
+          if (ar.succeeded()) {
+            iterable = ar.result();
+            if (canRead()) {
+              doRead();
+            }
+          } else {
+            close();
+            handleException(ar.cause());
+          }
+        }
+      });
+    }
+    return this;
+  }
+
+  private boolean canRead() {
+    return !paused && !closed;
+  }
+
+  @Override
+  public synchronized CloseableIteratorCollectionStream<IN, OUT> pause() {
+    checkClosed();
+    paused = true;
+    return this;
+  }
+
+  @Override
+  public synchronized CloseableIteratorCollectionStream<IN, OUT> resume() {
+    checkClosed();
+    if (paused) {
+      paused = false;
+      if (dataHandler != null) {
+        doRead();
+      }
+    }
+    return this;
+  }
+
+  private synchronized void doRead() {
+    if (readInProgress) {
+      return;
+    }
+    readInProgress = true;
+    if (iterator == null) {
+      context.<CloseableIterator<IN>>executeBlocking(fut -> fut.complete(iterable.iterator()), false, ar -> {
+        synchronized (this) {
+          readInProgress = false;
+          if (ar.succeeded()) {
+            iterator = ar.result();
+            if (canRead()) {
+              doRead();
+            }
+          } else {
+            close();
+            handleException(ar.cause());
+          }
+        }
+      });
+      return;
+    }
+    if (queue == null) {
+      queue = new ArrayDeque<>(BATCH_SIZE);
+    }
+    if (!queue.isEmpty()) {
+      context.runOnContext(v -> emitQueued());
+      return;
+    }
+    context.<List<IN>>executeBlocking(fut -> {
+      List<IN> batch = new ArrayList<>(BATCH_SIZE);
+      for (int i = 0; i < BATCH_SIZE && iterator.hasNext(); i++) {
+        batch.add(iterator.next());
+      }
+      fut.complete(batch);
+    }, false, ar -> {
+      synchronized (this) {
+        if (ar.succeeded()) {
+          queue.addAll(ar.result());
+          if (queue.isEmpty()) {
+            close();
+            if (endHandler != null) {
+              endHandler.handle(null);
+            }
+          } else {
+            emitQueued();
+          }
+        } else {
+          close();
+          handleException(ar.cause());
+        }
+      }
+    });
+  }
+
+  private void handleException(Throwable cause) {
+    if (exceptionHandler != null) {
+      exceptionHandler.handle(cause);
+    }
+  }
+
+  private synchronized void emitQueued() {
+    while (!queue.isEmpty() && canRead()) {
+      dataHandler.handle(converter.apply(queue.remove()));
+    }
+    readInProgress = false;
+    if (canRead()) {
+      doRead();
+    }
+  }
+
+  @Override
+  public synchronized CloseableIteratorCollectionStream<IN, OUT> endHandler(Handler<Void> handler) {
+    endHandler = handler;
+    return this;
+  }
+
+  private void close() {
+    closed = true;
+    context.executeBlocking(fut -> {
+      if (iterator != null) {
```

Comment:
iterator should be accessed under synchronization

Reply:
Done